### PR TITLE
rocAL - Video_pipeline Support

### PR DESCRIPTION
**Pull request by Indumathi31**
1. Video Reader to read, decode and return a batch of sequences in NFHWC format
2. Video Reader Resize to fuse reading and decoding
3. Sequence Reader to read sequence from directory of video stream
4. Sequence Rearrange to rearrange frames in sequence
5. Meta Data support for the video reader.
6. Enabled HIP Backend support for the video pipeline.

### CODE REVIEW

File: rocAL/rocAL/include/caffe_meta_data_reader.h [hunk 1]
```diff
@@ -38,6 +38,7 @@ public :
     void read_all(const std::string& path) override;
     void release(std::string image_name);
     void release() override;
+    void set_timestamp_mode() override { return; }
```
Comment (rrawther) [inline rocAL/rocAL/include/caffe_meta_data_reader.h:41]: if this function is not supported here, make it bool and return false

File: rocAL/rocAL/include/commons.h [hunk 2]
```diff
@@ -71,6 +71,16 @@ enum class RaliMemType
     HIP
 };
 
+/*! \brief Decoder mode for Video decoding
+ * 
+ *  Currently supports Software decoding, will support Hardware decoding in future
+ */
+enum class DecodeMode
+{
+    USE_HW = 0,
```
Comment (rrawther) [inline rocAL/rocAL/include/commons.h:80]: please rename USE_HW to HW_VAAPI and USE_SW to CPU

File: rocAL/rocAL/include/ffmpeg_video_decoder.h [hunk 3]
```diff
@@ -0,0 +1,48 @@
+/*
+Copyright (c) 2019 - 2020 Advanced Micro Devices, Inc. All rights reserved.
```
Comment (rrawther) [inline rocAL/rocAL/include/ffmpeg_video_decoder.h:2]: adjust the date to 2021

File: rocAL/rocAL/include/master_graph.h [hunk 4]
```diff
@@ -48,7 +50,7 @@ class MasterGraph
     MasterGraph(size_t batch_size, RaliAffinity affinity, int gpu_id, size_t cpu_threads, size_t prefetch_queue_depth, RaliTensorDataType output_tensor_data_type);
     ~MasterGraph();
     Status reset();
-    size_t remaining_images_count();
+    size_t remaining_images_or_sequences_count();
```
Comment (rrawther) [inline rocAL/rocAL/include/master_graph.h:53]: please rename this to remaining_count() instead.

File: rocAL/rocAL/include/master_graph.h [hunk 5]
```diff
@@ -136,7 +149,7 @@ class MasterGraph
     bool _first_run = true;
     bool _processing;//!< Indicates if internal processing thread should keep processing or not
     const static unsigned SAMPLE_SIZE = sizeof(unsigned char);
-    int _remaining_images_count;//!< Keeps the count of remaining images yet to be processed for the user,
+    int _remaining_images_or_sequences_count;//!< Keeps the count of remaining images yet to be processed for the user,
```
Comment (rrawther) [inline rocAL/rocAL/include/master_graph.h:152]: please rename it to _remaining_count() instead

File: .gitignore [hunk 6]
```diff
@@ -1,6 +1,4 @@
 build*
-docker_build*
```
Comment (rrawther) [inline .gitignore:2]: why is this file modified for this PR?

File: rocAL/rocAL/include/context.h [hunk 7]
```diff
@@ -53,6 +53,8 @@ struct Context
     }
     size_t user_batch_size() { return _user_batch_size; }
     size_t internal_batch_size() { return _internal_batch_size; }
+    void set_user_batch_size(size_t user_batch_size) {  _user_batch_size = user_batch_size; }
+    void set_internal_batch_size(size_t internal_batch_size) { _internal_batch_size = internal_batch_size; }
```
Comment (rrawther) [inline rocAL/rocAL/include/context.h:57]: In the previous discussion, I thought we have discussed to not to set internal_batch_size externally

File: rocAL/rocAL/include/sequence_file_source_reader.h [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+    /*!
+     \param desc  User provided descriptor containing the files' path.
+    */
+    Reader::Status initialize(ReaderConfig desc) override;
+    //! Reads the next resource item
+    /*!
+     \param buf User's provided buffer to receive the loaded images
+     \return Size of the loaded resource
+    */
+    size_t read(unsigned char* buf, size_t max_size) override;
```
Comment (rrawther) [inline rocAL/rocAL/include/sequence_file_source_reader.h:45]: todo: rename read() function name to avoid codacy warnings

File: rocAL/rocAL/include/text_file_meta_data_reader.h [hunk 9]
```diff
@@ -33,6 +33,7 @@ class TextFileMetaDataReader: public MetaDataReader
     void read_all(const std::string& path) override;
     void release(std::string image_name);
     void release() override;
+    void set_timestamp_mode() override { return; }
```
Comment (rrawther) [inline rocAL/rocAL/include/text_file_meta_data_reader.h:36]: avoid the use of null functions. If it is not supported please make it bool and return false

File: rocAL/rocAL/source/node_video_loader.cpp [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+    // Set reader and decoder config accordingly for the VideoLoaderNode
+    auto reader_cfg = VideoReaderConfig(storage_type, source_path, shuffle, loop);
+    reader_cfg.set_shard_count(internal_shard_count);
+    reader_cfg.set_batch_count(load_batch_count);
+    reader_cfg.set_sequence_length(sequence_length);
+    reader_cfg.set_frame_step(step);
+    reader_cfg.set_frame_stride(stride);
+    reader_cfg.set_video_properties(video_prop);
+    _loader_module->initialize(reader_cfg, VideoDecoderConfig(decoder_type),
+                               mem_type,
```
Comment (rrawther) [inline rocAL/rocAL/source/node_video_loader.cpp:56]: please combine ln#55, 56 and 57

**Comment by paveltc** [comment]
@rrawther This PR passes all unit tests specified in documentation.

File: rocAL/rocAL/source/rali_api_augmentation.cpp [hunk 11]
```diff
@@ -60,11 +60,48 @@ THE SOFTWARE.
...
+    Image* output = nullptr;
+    auto context = static_cast<Context*>(p_context);
+    try
+    {
+        if(sequence_length == 0)
+            THROW("sequence_length passed should be bigger than 0")
+        auto input = static_cast<Image*>(p_input);
+        auto info = ImageInfo(input->info().width(), input->info().height_single(),
+                              context->master_graph->internal_batch_size() * new_sequence_length,
+                              input->info().color_plane_count(),
```
Comment (rrawther) [inline rocAL/rocAL/source/rali_api_augmentation.cpp:91]: please follow the change is tot master to get rid of all codacy errors regarding !p_context

**Review by rrawther — CHANGES_REQUESTED** [review]
@Indumathi31 please merge with master and apply fix in the master for codacy errors

File: rocAL/rocAL/source/rali_api_augmentation.cpp (hunk 11 above)
Comment (Indumathi31) [reply rocAL/rocAL/source/rali_api_augmentation.cpp:92]: Have added the changes to fix codacy errors.

**Comment by Indumathi31** [comment]
@rrawther I have merged master and codacy errors were resolved after the fix.

**Comment by rrawther** [comment]
@paveltc: Can you please run tests on this PR
@LakshmiKumar23 : Can you run SSD training on this PR for performance?

**Comment by LakshmiKumar23** [comment]
> @paveltc: Can you please run tests on this PR @LakshmiKumar23 : Can you run SSD training on this PR for performance?

@rrawther SSD training converged in 59 minutes at the 55th epoch

**Comment by paveltc** [comment]
@rrawther I ran the tests on this PR and commented that they passed. I'm not sure what happened to my previous comment.

**Review by rrawther — APPROVED** [review]
@kiritigowda to check the Jenkins failure

**Comment by kiritigowda** [comment]
> @kiritigowda to check the Jenkins failure

passing CI
